### PR TITLE
Update docs on arg train_dataloader in fit

### PR DESCRIPTION
**Pull request by SkafteNicki**
## What does this PR do?

Fixes https://github.com/PyTorchLightning/pytorch-lightning/pull/1959#issuecomment-781763434
Updates docstring of `fit` to mention that `train_dataloader` now can take in multiple dataloaders.

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by rohitgr7** [comment]
maybe add some examples and update type hints here too: https://github.com/PyTorchLightning/pytorch-lightning/blob/0b271474e5cc35a40009559d6c4488ab3b02723e/pytorch_lightning/core/hooks.py#L386??

**Review by Borda — APPROVED** [review]

File: pytorch_lightning/core/hooks.py [hunk 1]
```diff
@@ -413,7 +415,7 @@ def train_dataloader(self) -> DataLoader:
             There is no need to set it yourself.
 
         Example::
-
+            # single dataloader
```
Comment (Borda) [inline pytorch_lightning/core/hooks.py:418]:
```suggestion

            # single dataloader
```

File: pytorch_lightning/core/hooks.py [hunk 2]
```diff
@@ -385,10 +385,12 @@ def prepare_data(self):
 
     def train_dataloader(self) -> DataLoader:
```
Comment (rohitgr7) [inline pytorch_lightning/core/hooks.py:386]:
```suggestion
    def train_dataloader(self) -> Any:
```

**Review by rohitgr7 — APPROVED** [review]

File: pytorch_lightning/core/hooks.py (hunk 2 above)
Comment (Borda) [reply pytorch_lightning/core/hooks.py:386]: why Any?

Comment (rohitgr7) [reply pytorch_lightning/core/hooks.py:386]: that's what the PR added here: https://github.com/PyTorchLightning/pytorch-lightning/pull/6076/files#diff-667a2513c158c2b01d73b479e4dea75587d8531f9cd286891d34570a2fd145dcR428

since `(list, dict, nested lists and dicts)` are supported for train_dataloader. although Any is too over simplified but we can be more specific here.

Comment (carmocca) [reply pytorch_lightning/core/hooks.py:386]: Do we document anywhere all those options? We should be clearer about it

**Review by justusschock — APPROVED** [review]

File: pytorch_lightning/trainer/trainer.py [hunk 3]
```diff
@@ -437,8 +437,9 @@ def fit(
 
             model: Model to fit.
 
-            train_dataloader: A Pytorch DataLoader with training samples. If the model has
-                a predefined train_dataloader method this will be skipped.
+            train_dataloader: Either a single Pytorch Dataloader or an collection of these
+                (list, dict, nested lists and dicts). In the case of multiple dataloaders, please
+                see this :ref:`page <multiple-training-dataloaders>`
```
Comment (awaelchli) [inline pytorch_lightning/trainer/trainer.py:442]:
```suggestion
            train_dataloader: Either a single PyTorch DataLoader or a collection of these
                (list, dict, nested lists and dicts). In the case of multiple dataloaders, please
                see this :ref:`page <multiple-training-dataloaders>`
```

File: pytorch_lightning/core/hooks.py [hunk 4]
```diff
@@ -426,6 +429,22 @@ def train_dataloader(self):
...
+                return [torch.utils.data.DataLoader(dataset=mnist, batch_size=self.batch_size, shuffle=True),
+                        torch.utils.data.DataLoader(dataset=cifar, batch_size=self.batch_size, shuffle=True)]
+
+            # multiple dataloader, return as dict
+            def train_dataloader(self):
+                mnist = MNIST(...)
+                cifar = CIFAR(...)
+                # each batch will be a dict of tensors: {'mnist': batch_mnist, 'cifar': batch_cifar}
+                return {'mnist': torch.utils.data.DataLoader(dataset=mnist, batch_size=self.batch_size, shuffle=True),
+                        'cifar': torch.utils.data.DataLoader(dataset=cifar, batch_size=self.batch_size, shuffle=True)]
```
Comment (awaelchli) [inline pytorch_lightning/core/hooks.py:446]:
```suggestion
                return {'mnist': torch.utils.data.DataLoader(dataset=mnist, batch_size=self.batch_size, shuffle=True),
                        'cifar': torch.utils.data.DataLoader(dataset=cifar, batch_size=self.batch_size, shuffle=True)}
```

File: docs/source/advanced/multiple_loaders.rst [hunk 5]
```diff
@@ -86,6 +88,27 @@ For more details please have a look at :attr:`~pytorch_lightning.trainer.trainer
 
             return loaders
 
+Furthermore, lightning also supports that nested lists and dicts (or an combination) can
+be returned
```
Comment (awaelchli) [inline docs/source/advanced/multiple_loaders.rst:92]:
```suggestion
Furthermore, Lightning also supports that nested lists and dicts (or a combination) can
be returned 
```

File: docs/source/advanced/multiple_loaders.rst [hunk 6]
```diff
@@ -86,6 +88,27 @@ For more details please have a look at :attr:`~pytorch_lightning.trainer.trainer
...
+            loader_a = torch.utils.data.DataLoader(range(8), batch_size=4)
+            loader_b = torch.utils.data.DataLoader(range(16), batch_size=4)
+            loader_c = torch.utils.data.DataLoader(range(32), batch_size=4)
+            loader_c = torch.utils.data.DataLoader(range(64), batch_size=4)
+
+            # pass loaders as a nested dict. This will create batches like this:
+            # {'loader_a_b': {'a': batch from loader a, 'b': batch from loader b},
+            #  'loader_c_d': {'c': batch from loader c, 'd': batch from loader d}}
+            loaders = {'loaders_a_b': {'a': loader_a, 'b': loader_b}
+                       'loaders_c_d': {'c': loader_c, 'd': loader_d}}
```
Comment (awaelchli) [inline docs/source/advanced/multiple_loaders.rst:109]:
```suggestion
            loaders = {'loaders_a_b': {'a': loader_a, 'b': loader_b},
                       'loaders_c_d': {'c': loader_c, 'd': loader_d}}
```
this is probably why the test fails :)

File: pytorch_lightning/core/hooks.py [hunk 7]
```diff
@@ -426,6 +429,22 @@ def train_dataloader(self):
                 )
                 return loader
 
+            # multiple dataloaders, return as list
+            def train_dataloader(self):
+                mnist = MNIST(...)
+                cifar = CIFAR(...)
+                # each batch will be a list of tensors: [batch_mnist, batch_cifar]
+                return [torch.utils.data.DataLoader(dataset=mnist, batch_size=self.batch_size, shuffle=True),
+                        torch.utils.data.DataLoader(dataset=cifar, batch_size=self.batch_size, shuffle=True)]
```
Comment (awaelchli) [inline pytorch_lightning/core/hooks.py:438]: Could we shorten or format so that the line length is a bit shorter? The generated docs wrap around which is less readable

image

Comment (SkafteNicki) [reply pytorch_lightning/core/hooks.py]: Done :]

**Review by tchaton — APPROVED** [review]

**Review by carmocca — APPROVED** [review]

**Review by s-rog — APPROVED** [review]

File: pytorch_lightning/core/hooks.py [hunk 8]
```diff
@@ -383,12 +383,14 @@ def prepare_data(self):
...
+    def train_dataloader(self) -> Any:
         """
-        Implement a PyTorch DataLoader for training.
+        Implement one or more PyTorch DataLoaders for training.
 
         Return:
-            Single PyTorch :class:`~torch.utils.data.DataLoader`.
+            Either a single PyTorch :class:`~torch.utils.data.DataLoader` or an collection of these
+            (list, dict, nested lists and dicts). In the case of multiple dataloaders, please see
+            this :ref:`page <multiple-training-dataloaders>`
```
Comment (awaelchli) [inline pytorch_lightning/core/hooks.py:393]:
```suggestion
            Either a single PyTorch :class:`~torch.utils.data.DataLoader` or a collection of these
            (list, dict, nested lists and dicts). In the case of multiple dataloaders, please see
            this :ref:`page <multiple-training-dataloaders>`
```

**Review by awaelchli — APPROVED** [review]
